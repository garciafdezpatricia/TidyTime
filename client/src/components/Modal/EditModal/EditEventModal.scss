@use "../../../../styles/variables" as *;

.edit-event-article {
    position: absolute;
	transform: translate(-50%, -50%);
	background-color: $white;
	border-radius: .4rem;
	box-sizing: border-box;
    z-index: 10;
    top: 50%;
	left: 50%;
	bottom: 50%;
	padding: 1rem 2rem;
	width: 35rem;
	height: 18rem;
    box-shadow: -.1rem .2rem 1.5rem 0 rgba(0,0,0,0.75);

    display: grid;
    grid-template-columns: 1fr 1fr;
    grid-template-rows: auto auto repeat(3, 1fr);
    font-size: .8rem;

    @media screen and (min-width: 320px) and (max-width: 768px) {
        width: 80%;
        grid-template-rows: auto auto repeat(3, 1fr) 2fr;
        height: 60vh;
    }

    .edit-event-header {
        grid-row: 1/1;
        grid-column: 1/-1;
        display: flex;
        gap: 1rem;
        width: fit-content;
        margin-left: auto;

        a.edit-event-header-button {
            display: flex;
        }

        button.edit-event-header-button, a.edit-event-header-button {
            border: none;
            width: fit-content;
            height: fit-content;
            padding: .3rem;
            border-radius: .4rem;
            background-color: $white;
            &:hover {
                background-color: $primary-grey;
            }
        }

        button.edit-event-close-btn {
            border: none;
            width: fit-content;
            height: fit-content;
            padding: .3rem;
            border-radius: .4rem;
            background-color: $primary-grey;
        }
    }

    .edit-event-title {
        grid-row: 2/2;
        grid-column: 1/1;
        font-size: 1.2rem;
        border: none;
        height: fit-content;
        justify-self: self-start;
        width: 100%;
        &:active, &:focus {
            border: none;
            outline: none;
            text-decoration: underline;
            text-decoration-color: $primary-green;
        }

        @media screen and (min-width: 320px) and (max-width: 768px) {
            grid-column: 1/-1;
        }
    }

    .edit-event-info {
        grid-row: 2/-1;
        grid-column: 2/2;
        resize: none;
        padding-left: .2rem;
        margin: .5rem;
        border-radius: .4rem;
        border: none;
        background-color: #e6e6e66a;
        color: $black;
        &:active, &:focus {
            border: none;
            outline: none;
        }

        @media screen and (min-width: 320px) and (max-width: 768px) {
            grid-row: 6/-1;
            grid-column: 1/-1;
            
        }
    }

    .edit-event-start-date {
        grid-row: 3/3;
        grid-column: 1/1;
        align-self: center;
    }

    .edit-event-end-date {
        grid-row: 4/4;
        grid-column: 1/1;
    }

    .edit-event-start-date, .edit-event-end-date {
        height: fit-content;
        padding-inline: .4rem;
        justify-self: self-start;
        width: 60%;

        @media screen and (min-width: 320px) and (max-width: 768px) {
            grid-column: 1/-1;
        }
    }

    .edit-event-color {
        grid-row: 5/-1;
        grid-column: 1/1;

        @media screen and (min-width: 320px) and (max-width: 768px) {
            grid-column: 1/-1;
        }
    }
}