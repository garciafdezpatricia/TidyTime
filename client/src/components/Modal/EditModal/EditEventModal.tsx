import { useEffect, useState } from "react";
import { RiDeleteBin6Line } from "react-icons/ri";
import { GrShareOption } from "react-icons/gr";
import ComboBox from "@/src/components/ComboBox/ComboBox";
import { IoMdClose } from "react-icons/io";
import { MdOutlineDone } from "react-icons/md";
import { TbEyeShare } from "react-icons/tb";
import { useEventContext } from "@/src/components/Context/EventContext";
import { Event } from "@/src/task/Scheme";
import { useClickAway } from "@uidotdev/usehooks";
import toast from "react-hot-toast";
import { PiWarningOctagonFill } from "react-icons/pi";
import ShareModal from "../ShareModal/ShareEventModal";
import PromptModal from "../PromptModal/PromptModal";

export interface Props {
    onClose: (arg?:any) => void | any;
}

export default function EditEventModal({onClose} : Props) {
    // event context utils
    const {setEvents, events, selectedEventId } = useEventContext();

    const [selectedIndex, setSelectedIndex] = useState(-1);
    const [newTitle, setNewTitle] = useState("");
    const [newDesc, setNewDesc] = useState("");
    const [newStartDate, setNewStartDate] = useState("");
    const [newEndDate, setNewEndDate] = useState("");
    const [color, setNewColor] = useState("");
    const [isGoogleEvent, setIsGoogleEvent] = useState(false);
    const [googleId, setGoogleId] = useState("");
    const [googleHtml, setGoogleHtml] = useState("");

    const [isConfirmationDeleteModalOpen, setConfirmationDeleteModalOpen] = useState(false);

    const ref = useClickAway(() => {
        onClose();
    })

    /**
     * Set the selected event data to correspondent states.
     */
    useEffect(() => {
        events.map((event, index) => {
            if (event.eventId === selectedEventId) {
                setSelectedIndex(index);
                setNewTitle(event.title);
                setNewDesc(event.desc);
                const startDate = new Date(event.start.getTime() - (event.start.getTimezoneOffset() * 60000)).toISOString().slice(0, 16);
                const endDate = new Date(event.end.getTime() - (event.end.getTimezoneOffset() * 60000)).toISOString().slice(0, 16);
                setNewStartDate(startDate);
                setNewEndDate(endDate);
                setNewColor(event.color ? event.color : "");
                if (event.googleId){
                    setIsGoogleEvent(true);
                    setGoogleId(event.googleId);
                    setGoogleHtml(event.googleHTML ?? "");
                }
                return;
            }
        })
    }, [events]);
    /**
     * Auxiliary function to compare original values of the event with the ones present when the saving-event action is triggered. 
     * @param eventToUpdate contains the event to be updated
     * @returns object with the values that have changed.
     */
    const checkChangedValues = (eventToUpdate: Event) => {
        // if any value is different from the original one, set the key of the changedValues
        // object to that new value.
        const changedValues = {title: "", desc: "", start: "", end: "", color: ""};
        if (eventToUpdate.title !== newTitle) {
            changedValues.title = newTitle;
        }
        if (eventToUpdate.desc !== newDesc) {
            changedValues.desc = newDesc;
        }
        // format date to YYYY-MM-DDThh:mm:ss in correct time zone
        const startDate = new Date(eventToUpdate.start.getTime() - (eventToUpdate.start.getTimezoneOffset() * 60000)).toISOString().slice(0, 16);
        const endDate = new Date(eventToUpdate.end.getTime() - (eventToUpdate.end.getTimezoneOffset() * 60000)).toISOString().slice(0, 16);

        if (startDate !== newStartDate) {
            changedValues.start = newStartDate;
        }
        if (endDate !== newEndDate) {
            changedValues.end = newEndDate;
        }
        if (eventToUpdate.color !== color) {
            changedValues.color = color;
        }
        // check if changedValues has any new value
        const allValuesEmpty = Object.values(changedValues).every(value => value === "");
        // return null if changedValues does not have any new value
        return allValuesEmpty ? null : changedValues;
    }

    /**
     * Only need to update google event if title, desc, start date or end date changed.
     * @param valuesToChange 
     * @returns true if the google update is needed. False otherwise.
     */
    const needToUpdateGoogle = (valuesToChange:{title: string, desc: string, start: string, end: string, color: string}) => {
        // color changes do not have an influence on google update
        return valuesToChange.title !== "" || valuesToChange.desc !== "" || valuesToChange.start !== "" || valuesToChange.end !== "";

    }

    /**
     * Updates event based on the values that changed
     * @param event the event to be updated
     * @param valuesToChange the values of the event that need to be updated
     */
    const updateEvent = (event:Event, valuesToChange:{title: string, desc: string, start: string, end: string, color: string}) => {
        if (valuesToChange.title !== "") {
            event.title = valuesToChange.title;
        }
        if (valuesToChange.desc !== "") {
            event.desc = valuesToChange.desc;
        }
        if (valuesToChange.start !== "") {
            event.start = new Date(valuesToChange.start);
        }
        if (valuesToChange.end !== "") {
            event.end = new Date(valuesToChange.end);
        }
        if (valuesToChange.color !== "") {
            event.color = valuesToChange.color;
        }
    }

    /**
     * Update event on Google Calendar through API call.
     * @param eventToUpdate the event to be updated 
     */
    const updateEventOnGoogle = (eventToUpdate:Event) => {
        // format date to google format
        const ISOStartDate = new Date(new Date(eventToUpdate.start).getTime() - (new Date(eventToUpdate.start).getTimezoneOffset() * 60000)).toISOString();
        const ISOEndDate = new Date(new Date(eventToUpdate.end).getTime() - (new Date(eventToUpdate.end).getTimezoneOffset() * 60000)).toISOString();

        return fetch('http://localhost:8080/google/events/update', {
            method: 'POST',
            headers: {
            'Content-Type': 'application/json',
            },
            body: JSON.stringify({ 
                id: googleId,
                start: ISOStartDate,
                end: ISOEndDate,
                title: eventToUpdate.title,
                desc: eventToUpdate.desc,
                calendarId: eventToUpdate.googleCalendar                     
            }),
            credentials: 'include',
        })
        .then(response => response.json())
        .then(data => {
            if (data.status === 'error'){
                throw data.value;
            }
        })
        .catch(error => {
            throw error;
        });
    }

    /**
     * Save changes made to the event.
     */
    const onSave = () => {
        // get event to be saved
        let updatedEvents = [...events];
        const eventToUpdate = events[selectedIndex];
        // check which information changed
        const changedValues = checkChangedValues(eventToUpdate);
        if (changedValues) {
            // save correspondent information
            updateEvent(eventToUpdate, changedValues);
            updatedEvents = [
                ...events.slice(0, selectedIndex),
                eventToUpdate,
                ...events.slice(selectedIndex + 1)
            ];
            setEvents(updatedEvents);
            toast.success('Saved event!', {
                position: "bottom-center"
            });
            // google flow
            if (isGoogleEvent) {
                if (needToUpdateGoogle(changedValues)) {
                    const promise = updateEventOnGoogle(eventToUpdate);
                    toast.promise(promise, {
                        loading: "Updating event...",
                        success: "Google is updated!", 
                        error: (err) => "Google says '" + err + "'"
                    }, {
                        position: "bottom-center",
                        error: {
                            icon: <PiWarningOctagonFill />
                        }
                    });
                }
            }
        }
        // close edit modal
        onClose();
    }

    /**
     * Handler for color changes.
     * @param newColor new color of the event.
     */
    const handleColorChange = (newColor:string) => {
        setNewColor(newColor);
    }

    /**
     * Deletes the event from the set of events.
     */
    const onDelete = () => {
        setEvents((prevEvents) => {
			return prevEvents.filter((_, i) => i !== selectedIndex);
		});
        toast.success('Event succesfully deleted!', {position: "bottom-center"});
        // close edit modal
        onClose();
    }

    /**
     * Exports event to Google Calendar.
     */
    const onShare = () => {
        
    }

    return (
        // @ts-ignore
        <article ref={ref} className="edit-event-article">
            <section className="edit-event-header">
                {isGoogleEvent && 
                    <a className="edit-event-header-button" title="See in Google" href={googleHtml} target="_blank" rel="noopener noreferrer">
                        <TbEyeShare color="#363535" size={"1.2rem"} />
                    </a>
                }
                <button className="edit-event-header-button" title="Save" onClick={() => onSave()}>
                    <MdOutlineDone color="#363535" size={"1.2rem"} />
                </button>
                <button className="edit-event-header-button" title="Delete" onClick={() => setConfirmationDeleteModalOpen(true)}>
                    <RiDeleteBin6Line color="#363535" size={"1.1rem"} />
                </button>
                { !isGoogleEvent &&
                    <ShareModal selectedIndex={selectedIndex} />
                }
                <button className="edit-event-close-btn" onClick={() => onClose()}>
                    <IoMdClose size={"1.2rem"} />
                </button>
            </section>
            <input 
                placeholder="New title..." 
                className="edit-event-title" 
                type="text" 
                value={newTitle} 
                onChange={(e) => setNewTitle(e.target.value)} 
            />
            <textarea 
                placeholder="Add notes..." 
                className="edit-event-info" 
                value={newDesc}
                onChange={(e) => setNewDesc(e.target.value)} 
            />
            <input 
                className="edit-event-start-date" 
                type="datetime-local"
                defaultValue={newStartDate}
                onChange={(e) => setNewStartDate(e.target.value)}
            />
            <input 
                className="edit-event-end-date" 
                type="datetime-local"
                defaultValue={newEndDate}
                onChange={(e) => setNewEndDate(e.target.value)}
            />
            <div className="edit-event-color">
                <ComboBox
                    onOptionChange={handleColorChange}
                    text={"Color:"} 
                    colors={["#e13535", "#d37373", "#ffa500", "#3E5B41", "#3bb4ff"]}
                    checkedOption={ color !== "" ? color : "#3E5B41"} 
                />
            </div>
            {isConfirmationDeleteModalOpen && (
				<PromptModal
					title="Are you sure you want to delete this list? This action can't be undone"
					onPrimaryAction={() => onDelete()}
					primaryActionText='Delete'
					secondaryActionText='Cancel'
					onSecondaryAction={() => setConfirmationDeleteModalOpen(false)}
					variant='confirmation-modal'
					backdrop
				></PromptModal>
			)}
        </article>
    )
}