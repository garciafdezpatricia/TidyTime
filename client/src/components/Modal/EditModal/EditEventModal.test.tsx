import { render } from "@testing-library/react";
import '@testing-library/jest-dom';
import EditEventModal from '@/src/components/Modal/EditModal/EditEventModal';

describe('EditEventModal', () => {
    it('renders without crashing', () => {
        const { container } = render(
        <EditEventModal onClose={function (arg?: any) {
                throw new Error("Function not implemented.");
            } }   />);
        expect(container).toBeInTheDocument();
    });

    it('matches snapshot', () => {
        const { asFragment } = render(
        <EditEventModal onClose={function (arg?: any) {
                throw new Error("Function not implemented.");
            } }   />);
        expect(asFragment()).toMatchSnapshot();
      });
  });