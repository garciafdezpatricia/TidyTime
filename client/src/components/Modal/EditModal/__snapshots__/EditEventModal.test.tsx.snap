// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`EditEventModal matches snapshot 1`] = `
<DocumentFragment>
  <article
    class="edit-event-article"
  >
    <section
      class="edit-event-header"
    >
      <button
        class="edit-event-header-button"
        title="calendar.eventPanel.buttons.save"
      >
        <svg
          color="#363535"
          fill="currentColor"
          height="1.2rem"
          stroke="currentColor"
          stroke-width="0"
          style="color: rgb(54, 53, 53);"
          viewBox="0 0 24 24"
          width="1.2rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M0 0h24v24H0V0z"
            fill="none"
          />
          <path
            d="M9 16.2 4.8 12l-1.4 1.4L9 19 21 7l-1.4-1.4L9 16.2z"
          />
        </svg>
      </button>
      <button
        class="edit-event-header-button"
        title="calendar.eventPanel.buttons.delete"
      >
        <svg
          color="#363535"
          fill="currentColor"
          height="1.1rem"
          stroke="currentColor"
          stroke-width="0"
          style="color: rgb(54, 53, 53);"
          viewBox="0 0 24 24"
          width="1.1rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M7 4V2H17V4H22V6H20V21C20 21.5523 19.5523 22 19 22H5C4.44772 22 4 21.5523 4 21V6H2V4H7ZM6 6V20H18V6H6ZM9 9H11V17H9V9ZM13 9H15V17H13V9Z"
          />
        </svg>
      </button>
      <div
        class="button-share"
      >
        <button
          class="edit-event-header-button"
          title="calendar.eventPanel.buttons.share"
        >
          <svg
            color="#363535"
            fill="currentColor"
            height="1.1rem"
            stroke="currentColor"
            stroke-width="0"
            style="color: rgb(54, 53, 53);"
            viewBox="0 0 24 24"
            width="1.1rem"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M18,8 C19.6568542,8 21,6.65685425 21,5 C21,3.34314575 19.6568542,2 18,2 C16.3431458,2 15,3.34314575 15,5 C15,6.65685425 16.3431458,8 18,8 Z M6,15 C7.65685425,15 9,13.6568542 9,12 C9,10.3431458 7.65685425,9 6,9 C4.34314575,9 3,10.3431458 3,12 C3,13.6568542 4.34314575,15 6,15 Z M18,22 C19.6568542,22 21,20.6568542 21,19 C21,17.3431458 19.6568542,16 18,16 C16.3431458,16 15,17.3431458 15,19 C15,20.6568542 16.3431458,22 18,22 Z M16,18 L8,13 M16,6 L8,11"
              fill="none"
              stroke-width="2"
            />
          </svg>
        </button>
      </div>
      <button
        class="edit-event-close-btn"
      >
        <svg
          fill="currentColor"
          height="1.2rem"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 512 512"
          width="1.2rem"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M405 136.798L375.202 107 256 226.202 136.798 107 107 136.798 226.202 256 107 375.202 136.798 405 256 285.798 375.202 405 405 375.202 285.798 256z"
          />
        </svg>
      </button>
    </section>
    <input
      class="edit-event-title"
      placeholder="calendar.eventPanel.title"
      type="text"
      value=""
    />
    <textarea
      class="edit-event-info"
      placeholder="calendar.eventPanel.notes"
    />
    <input
      class="edit-event-start-date"
      type="datetime-local"
      value=""
    />
    <input
      class="edit-event-end-date"
      type="datetime-local"
      value=""
    />
    <div
      class="edit-event-color"
    >
      <div
        class="combobox"
      >
        <div
          class="combobox-header"
        >
          <label
            class="combobox-selected-color"
            style="background-color: rgb(62, 91, 65);"
          />
          <svg
            color="#000000"
            cursor="pointer"
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            stroke-width="0"
            style="color: rgb(0, 0, 0);"
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M48 256c0 114.9 93.1 208 208 208s208-93.1 208-208S370.9 48 256 48 48 141.1 48 256zm289.1-43.4c7.5-7.5 19.8-7.5 27.3 0 3.8 3.8 5.6 8.7 5.6 13.6s-1.9 9.9-5.7 13.7l-94.3 94c-7.6 6.9-19.3 6.7-26.6-.6l-95.7-95.4c-7.5-7.5-7.6-19.7 0-27.3 7.5-7.5 19.7-7.6 27.3 0l81.1 81.9 81-79.9z"
            />
          </svg>
        </div>
      </div>
    </div>
  </article>
</DocumentFragment>
`;
